staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -109,21 +109,140 @@ type unstructuredConverter struct {
 // test tools. Use DefaultUnstructuredConverter if you do not explicitly need mismatch detection.
 func NewTestUnstructuredConverter(comparison conversion.Equalities) UnstructuredConverter {
+	return NewTestUnstructuredConverterWithValidation(comparison)
+}
+
+// NewTestUnstrucutredConverterWithValidation allows for access to
+// FromUnstructuredWithValidation from within tests.
+func NewTestUnstructuredConverterWithValidation(comparison conversion.Equalities) *unstructuredConverter {
 	return &unstructuredConverter{
 		mismatchDetection: true,
 		comparison:        comparison,
 	}
 }
 
-// FromUnstructured converts an object from map[string]interface{} representation into a concrete type.
+// fromUnstructuredContext provides options for informing the converter
+// the state of its recursive walk through the conversion process.
+type fromUnstructuredContext struct {
+	// isInlined indicates whether the converter is currently in
+	// an inlined field or not to dtermine whether it should
+	// validate the matchedKeys yet or only collect them.
+	isInlined bool
+	// matchedKeys is a stack of the set of all fields that exist in the
+	// concrete go type of the object being converted into.
+	matchedKeys []map[string]struct{}
+	// parentPath collects the path that the conversion
+	// takes as it travers the unstructured json map.
+	// It is used to report the full path to any unknown
+	// fields that the converter encounters.
+	parentPath []string
+	// returnUnknownFields indicates whether or not
+	// unknown field errors should be collected and
+	// returned to the caller
+	returnUnknownFields bool
+	// unknownFieldErrors are the collection of
+	// the pull path to each unknown field in the
+	// object.
+	unknownFieldErrors []error
+}
+
+// pushMatchedKeyTracker adds a placeholder set for tracking
+// matched keys for the given level
+func (c *fromUnstructuredContext) pushMatchedKeyTracker() {
+	if !c.returnUnknownFields {
+		return
+	}
+	if !c.isInlined {
+		c.matchedKeys = append(c.matchedKeys, nil)
+	}
+}
+
+// recordMatchedKey initializes the last element of matchedKeys
+// (if needed) and sets 'key'
+func (c *fromUnstructuredContext) recordMatchedKey(key string) {
+	if !c.returnUnknownFields {
+		return
+	}
+	last := len(c.matchedKeys) - 1
+	if c.matchedKeys[last] == nil {
+		c.matchedKeys[last] = map[string]struct{}{}
+	}
+	c.matchedKeys[last][key] = struct{}{}
+}
+
+// verifyMatchedKeys pops the last element of matchedKeys,
+// checks the matched keys against the data, and adds unknown
+// field errors for any matched keys.
+// `data` are all of the keys that exist is the sv at this level (sv.MapKeys)
+// `matchedKeys` are all the keys found for that level
+func (c *fromUnstructuredContext) verifyMatchedKeys(data []reflect.Value) {
+	if !c.returnUnknownFields || c.isInlined {
+		return
+	}
+
+	var curMatchedKeys map[string]struct{}
+	last := len(c.matchedKeys) - 1
+	curMatchedKeys, c.matchedKeys = c.matchedKeys[last], c.matchedKeys[:last]
```

Comment:
need to nil out the element in matchedKeys before shortening the slice to avoid a retained reference to the map

```
curMatchedKeys := c.matchedKeys[last]
c.matchedKeys[last] = nil
c.matchedKeys = c.matchedKeys[:last]
```